File: internal/model/file.go
package model

import (
	"time"
)

// File 文件模型
type File struct {
	TenantBaseModel
	OriginalName string `json:"original_name" gorm:"not null;size:255" validate:"required,max=255"`
	FileName     string `json:"file_name" gorm:"not null;size:255" validate:"required,max=255"`
	FilePath     string `json:"file_path" gorm:"not null;size:500" validate:"required,max=500"`
	FileSize     int64  `json:"file_size" gorm:"not null" validate:"required,min=0"`
	FileType     string `json:"file_type" gorm:"not null;size:100" validate:"required,max=100"`
	MimeType     string `json:"mime_type" gorm:"not null;size:100" validate:"required,max=100"`
	MD5          string `json:"md5" gorm:"not null;size:32;index:idx_md5" validate:"required,len=32"`
	UploadUserID uint64 `json:"upload_user_id" gorm:"not null;index:idx_upload_user_id" validate:"required"`
	UsageType    string `json:"usage_type" gorm:"size:50;index:idx_usage_type" validate:"max=50"`

	// 关联关系
	UploadUser *User `json:"upload_user,omitempty" gorm:"foreignKey:UploadUserID"`
}

// TableName 指定表名
func (File) TableName() string {
	return "files"
}

// FileProfile 文件资料（简化版本）
type FileProfile struct {
	ID           uint64    `json:"id"`
	TenantID     uint64    `json:"tenant_id"`
	OriginalName string    `json:"original_name"`
	FileName     string    `json:"file_name"`
	FilePath     string    `json:"file_path"`
	FileSize     int64     `json:"file_size"`
	FileType     string    `json:"file_type"`
	MimeType     string    `json:"mime_type"`
	MD5          string    `json:"md5"`
	UploadUserID uint64    `json:"upload_user_id"`
	UsageType    string    `json:"usage_type"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
}

// ToProfile 转换为文件资料
func (f *File) ToProfile() FileProfile {
	return FileProfile{
		ID:           f.ID,
		TenantID:     f.TenantID,
		OriginalName: f.OriginalName,
		FileName:     f.FileName,
		FilePath:     f.FilePath,
		FileSize:     f.FileSize,
		FileType:     f.FileType,
		MimeType:     f.MimeType,
		MD5:          f.MD5,
		UploadUserID: f.UploadUserID,
		UsageType:    f.UsageType,
		CreatedAt:    f.CreatedAt,
		UpdatedAt:    f.UpdatedAt,
	}
}

// GetSizeInKB 获取文件大小（KB）
func (f *File) GetSizeInKB() float64 {
	return float64(f.FileSize) / 1024
}

// GetSizeInMB 获取文件大小（MB）
func (f *File) GetSizeInMB() float64 {
	return float64(f.FileSize) / (1024 * 1024)
}

// IsImage 检查是否为图片文件
func (f *File) IsImage() bool {
	imageTypes := []string{"image/jpeg", "image/png", "image/gif", "image/webp", "image/svg+xml"}
	for _, imgType := range imageTypes {
		if f.MimeType == imgType {
			return true
		}
	}
	return false
}

// IsDocument 检查是否为文档文件
func (f *File) IsDocument() bool {
	docTypes := []string{
		"application/pdf",
		"application/msword",
		"application/vnd.openxmlformats-officedocument.wordprocessingml.document",
		"application/vnd.ms-excel",
		"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
		"text/plain",
	}
	for _, docType := range docTypes {
		if f.MimeType == docType {
			return true
		}
	}
	return false
}

File: web/src/api/file.ts
import request from '@/utils/request'

export interface FileProfile {
  id: number
  tenantId: number
  originalName: string
  fileName: string
  filePath: string
  fileSize: number
  fileType: string
  mimeType: string
  md5: string
  uploadUserId: number
  usageType: string
  storageType: string
  isPublic: boolean
  accessUrl: string
  createdAt: string
  updatedAt: string
}

export interface FileListResponse {
  files: FileProfile[]
  pagination: {
    page: number
    pageSize: number
    total: number
  }
}

// 上传文件
export const uploadFile = (formData: FormData) => {
  return request({
    url: '/v1/files/upload',
    method: 'post',
    data: formData,
    headers: {
      'Content-Type': 'multipart/form-data'
    }
  })
}

// 获取文件信息（包含access_url用于下载、预览、复制链接）
export const getFile = (id: number) => {
  return request({
    url: `/v1/files/${id}`,
    method: 'get'
  })
}

// 使用access_url进行文件下载
export const downloadFileByUrl = (accessUrl: string, originalName: string) => {
  const link = document.createElement('a')
  link.href = accessUrl
  link.download = originalName
  link.style.display = 'none'
  document.body.appendChild(link)
  
  // 触发下载
  setTimeout(() => {
    link.click()
    console.log('Download triggered for:', originalName)
  }, 10)
  
  // 清理DOM
  setTimeout(() => {
    if (document.body.contains(link)) {
      document.body.removeChild(link)
    }
  }, 1000)
}

// 复制文件链接到剪贴板
export const copyFileUrl = async (accessUrl: string) => {
  try {
    await navigator.clipboard.writeText(accessUrl)
    return true
  } catch (err) {
    console.error('复制链接失败:', err)
    // 降级方案：使用传统方法
    const textArea = document.createElement('textarea')
    textArea.value = accessUrl
    textArea.style.position = 'fixed'
    textArea.style.left = '-999999px'
    textArea.style.top = '-999999px'
    document.body.appendChild(textArea)
    textArea.focus()
    textArea.select()
    
    try {
      document.execCommand('copy')
      document.body.removeChild(textArea)
      return true
    } catch (err) {
      document.body.removeChild(textArea)
      return false
    }
  }
}

// 预览图片文件（直接使用access_url）
export const previewImage = (accessUrl: string) => {
  return accessUrl // 直接返回URL，用于img标签的src
}

// 根据MIME类型获取文件扩展名
const getExtensionFromMimeType = (mimeType: string): string => {
  const mimeToExt: Record<string, string> = {
    // 图片类型
    'image/jpeg': '.jpg',
    'image/jpg': '.jpg',
    'image/png': '.png',
    'image/gif': '.gif',
    'image/webp': '.webp',
    'image/svg+xml': '.svg',
    'image/bmp': '.bmp',
    'image/tiff': '.tiff',

    // 文档类型
    'application/pdf': '.pdf',
    'application/msword': '.doc',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document': '.docx',
    'application/vnd.ms-excel': '.xls',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet': '.xlsx',
    'application/vnd.ms-powerpoint': '.ppt',
    'application/vnd.openxmlformats-officedocument.presentationml.presentation': '.pptx',

    // 文本类型
    'text/plain': '.txt',
    'text/html': '.html',
    'text/css': '.css',
    'text/javascript': '.js',
    'application/json': '.json',
    'text/xml': '.xml',
    'application/xml': '.xml',

    // 压缩文件
    'application/zip': '.zip',
    'application/x-rar-compressed': '.rar',
    'application/x-7z-compressed': '.7z',
    'application/x-tar': '.tar',
    'application/gzip': '.gz',

    // 音频
    'audio/mpeg': '.mp3',
    'audio/wav': '.wav',
    'audio/ogg': '.ogg',
    'audio/mp4': '.m4a',

    // 视频
    'video/mp4': '.mp4',
    'video/avi': '.avi',
    'video/mov': '.mov',
    'video/wmv': '.wmv',
    'video/webm': '.webm'
  }

  return mimeToExt[mimeType] || ''
}

// 删除文件
export const deleteFile = (id: number) => {
  return request({
    url: `/v1/files/${id}`,
    method: 'delete'
  })
}

// 获取用户文件列表
export const getUserFiles = (params: {
  page?: number
  pageSize?: number
  filename?: string
  fileType?: string
  usageType?: string
}) => {
  return request({
    url: '/v1/files',
    method: 'get',
    params
  })
}

// 获取所有文件列表（管理员）
export const getAllFiles = (params: {
  page?: number
  pageSize?: number
  filename?: string
  fileType?: string
  usageType?: string
  uploadUserId?: number
}) => {
  return request({
    url: '/v1/files',
    method: 'get',
    params
  })
}